### ISSUE TITLE
Mono Support

### ISSUE THREAD

**Issue by jaddie** [gitextensions/gitextensions#6051]
I wanted to create this issue to ask why mono support is being removed? Surely once its working its far easier to keep it going than to end up moving away from it which will make the future effort an even bigger chore?

Also on the note of remaining cross platform, is there any consideration towards using .net core instead as this might be a more future proof and suitable platform to work with?

I re-wrote something as large as WCell to work with Mono so might be able to lend a hand in places with GE since I use it as my main git client (albeit on Windows atm) as I like its fast reactions and no-nonsense approach, unlike Sourcetree that seem intent on making fancy animations and filling the client with crap that bogs it down.

**Comment by vbjay** [comment]
See #5914 #5940

**Comment by gerhardol** [comment]
Mono was removed in #4415, discussion in #4308 
Basically, WinForms support in Mono put a lot of limitations and there were very few contributions from Mono users so Mono was slowing down development considerably.

.NET Core 3 will hopefully happen (there are contributors that are working with .NET Core in other projects). The limitation is still WinForms. Microsoft has open sourced WinForms recently, so if someone ports WF to other platforms, multiplatform may emerge again.

**Comment by RussKie** [comment]
I believe @sharwell and @drewnoakes started discussions about the migration (e.g. https://github.com/gitextensions/Git.hub/pull/9). 
But I don't expect this journey to be quick or quirk free.

**Comment by RussKie** [comment]
> might be able to lend a hand in places

Help is always appreciated :+1:

**Comment by Mart-Bogdan** [comment]
Ah that's sad. One of the best Git client present on Linux.

Netcore 3 won't help, unless it's rewritten on Avalonia, but it seems unlikely,

**Comment by mletterle** [comment]
Just discovered this today, and it is indeed sad. GitExtensions is one of my favorite repo visualizers. Understandable though. Cross-Platform UI is becoming a bigger deal, hopefully we'll see some solution emerge.

**Comment by Wirone** [comment]
Side question: is it possible to run GitExtensions v3 with Wine?

**Comment by RussKie** [comment]
You could try, but I highly doubt it.
v2.51 is still available for the Linux users.

**Comment by Wirone** [comment]
@RussKie yeah, I know, I use it every day :wink: But it would be good to use newest one.

**Comment by arisliang** [comment]
It seems WinForms is in the progress of being ported to dotnet core 3 (https://github.com/dotnet/winforms). Does this mean we're going to have x-platform GitExtensions based on dotnet core 3 soon after?

**Comment by pmiossec** [comment]
@arisliang unfortunately, being ported to dotnet core doesn't mean that it will be cross-platform.

So the answer is no.

**Comment by gerhardol** [comment]
Active Linux/mono users could maintain the 2.x branch too.
This was the major reason to why Mono support had to be dropped, there was too much overhead.

**Comment by jaddie** [comment]
> @arisliang unfortunately, being ported to dotnet core doesn't mean that it will be cross-platform.
> 
> So the answer is no.

Why is this? Since core 3.x is going to have the UI side of things brought in, this would mean that git extensions could be supported this way on other platforms instead surely?

When I have some free time I'll see if I can start having a look at the comments about this and see how hard it might be to convert or not as was already discussing some things for other open source projects with some MS side people :)

**Comment by jaddie** [comment]
Nevermind, apparently some of the conversations I'd had were information that hasn't come to light properly yet and core only supports Windows for UI as its built on windows system libraries at this time, a rather stupid stop-gap considering things like Avalonia exist already. Might see if a thought of integrating this into the MS stack could happen in the future :D

**Comment by RussKie** [comment]
Unfortunately there are no immediate (even speculative) plans for WinForms or WPF .NET Core to become available on non Windows platforms.

**Personally**, I'd like to think by the time NET5 becomes a reality there will be someone adventurous enough to make inroads into the x-plat story.

**Comment by brunoais** [comment]
Not your fault.... It just seems to be a failed dream....

**Comment by mkbel** [comment]
> Side question: is it possible to run GitExtensions v3 with Wine?

I was able to successfully install version 3.1.1. Later versions complains about missing Windows PowerShell. But it depends on Git, and I didn't manage to install Git for windows. I also tried to use Linux native Git form GitExtensions inside Wine. This would be even more preferable way. But it looks like GitExtensions is unable get output from linux native git. I see output from git call inside console which runs wine, but that output or return status is probably not captured inside GitExtensions.

**Comment by gerhardol** [comment]
> Later versions complains about missing Windows PowerShell. 

Try the  portable version, this script is used for checking Telemetry

**Comment by mkbel** [comment]
Yes, portable version 3.3.1.7897 can be started in wine. But this was not the biggest issue. Biggest issue is, that git is needed and it doesn't run in wine. Your idea about trying portable version made me to try also git portable under wine, but it is crashing same as version with installer. To me only viable option would be to use native linux git from wine. It is probably not working because of this https://bugs.winehq.org/show_bug.cgi?id=18335 and it is closed as WONTFIX. So some workaround in GitExtensions or some wrapper for linux git to be able to return its status inside wine would be needed. Also I don't know, what is the first git command GitExtensions is trying to call git with. I have suspicion that also command line argument passing doesn't work, as I see this git output in console:

> wine GitExtensions.exe
> 0030:err:ole:CoGetContextToken apartment not initialised
> 0009:err:combase:RoGetActivationFactory Failed to find library for L"Windows.Foundation.Diagnostics.AsyncCausalityTracer"
> usage: git [--version] [--help] [-C <path>] [-c <name>=<value>]
>            [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
>            [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
>            [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
>            <command> [<args>]
> 
> These are common Git commands used in various situations:
> 
> start a working area (see also: git help tutorial)
>    clone      Clone a repository into a new directory
>    init       Create an empty Git repository or reinitialize an existing one
> 
> work on the current change (see also: git help everyday)
>    add        Add file contents to the index
>    mv         Move or rename a file, a directory, or a symlink
>    reset      Reset current HEAD to the specified state
>    rm         Remove files from the working tree and from the index
> 
> examine the history and state (see also: git help revisions)
>    bisect     Use binary search to find the commit that introduced a bug
>    grep       Print lines matching a pattern
>    log        Show commit logs
>    show       Show various types of objects
>    status     Show the working tree status
> 
> grow, mark and tweak your common history
>    branch     List, create, or delete branches
>    checkout    Switch branches or restore working tree files
>    commit     Record changes to the repository
>    diff       Show changes between commits, commit and working tree, etc
>    merge      Join two or more development histories together
>    rebase     Reapply commits on top of another base tip
>    tag        Create, list, delete or verify a tag object signed with GPG
> 
> collaborate (see also: git help workflows)
>    fetch      Download objects and refs from another repository
>    pull       Fetch from and integrate with another repository or a local branch
>    push       Update remote refs along with associated objects
> 
> 'git help -a' and 'git help -g' list available subcommands and some
> concept guides. See 'git help <command>' or 'git help <concept>'
> to read about a specific subcommand or concept.

I would expect GitExtensions to call something like `git --version` and not calling git without arguments.

**Comment by mstv** [comment]
These are the git commands called until the dashboard is shown on Win
(taken from the GE git command log - hotkey `F12`):
```
git	config --get ą
git
git	--version
git	--no-optional-locks for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"
git	--no-optional-locks show-ref --dereference
git	rev-parse --git-common-dir
git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/
git	remote -v
git	symbolic-ref HEAD
git	show-ref --dereference
git	remote
```

The git executable to be used, can be specified in `GitExtensions.settings`.
```
  <item>
    <key>
      <string>gitcommand</string>
    </key>
    <value>
      <string>git</string>
    </value>
  </item>
```

**Comment by mkbel** [comment]
So the installed version 3.1.1 is slightly more successful than portable 3.3.1 under wine.
For 3.1.1 I see output of first two git calls:

> error: key does not contain a section: ą

> usage: git [--version] [--help] [-C <path>] [-c <name>=<value>]
>            [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
>            [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
>            [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
>            <command> [<args>]

For 3.3.1 portable I see output of only second one.

Then booth versions open error window complaining about git not found.

![image](https://user-images.githubusercontent.com/35807926/80040454-6803ee80-84fa-11ea-8f30-4dc527b83356.png)

So because of seeing the error: key does not contain a section: ą we can clearly assert, that GitExtensions is successfully passing command line arguments to git.

So the only problem is, that it is not getting output and/or return value of git call.

**Comment by mstv** [comment]
> So the only problem is, that it is not getting output and/or return value of git call.

I think this will be difficult because `CreateProcess` does not return a valid handle. That's why GE cannot wait for git to terminate and cannot access its output.
A WINE executable seems to be needed which wraps the call of the native git.

**Comment by mkbel** [comment]
So I managed to workaround half of the problem.
Create symbolic link with .exe extension pointing to your git
`ln -s /usr/bin/git git.exe`
create batch file git.bat which calls your git.exe:
`/usr/bin/git.exe %*`
Point GitExtensions to use git.bat as git.

This solved the problem that GE cannot wait for git process with invalid handle to terminate.

But it still complains, that it can not parse its output.

I am working on getting also output of git command to GitExtension but haven't succeeded yet.
I managed to get output of git to wineconsole, but it is somehow not working in GitExtensions.
git.bat content:
```
REM works in wineconsole but not with gitextensions
@ECHO OFF
git.exe %*
:start
IF EXIST "lock.txt" (
    TYPE out.txt
    DEL out.txt
    DEL lock.txt
) ELSE (
    ping -n 1
    GOTO start
)
```
git.exe content:
```
#!/bin/bash
git "$@" > out.txt
touch lock.txt
```

**Comment by mkbel** [comment]
Managed to get some linux git output to wine GitExtensions:
Just changed paths to temporary files in batch and shell script to absolute and renamed git.exe to worker.exe to be sure it is not called directly. Still it is little bit broken - probably file based mutex is broken. Looks like GitExtensions is calling more git processes in parallel.
git.bat content:
```
@ECHO OFF
worker.exe %*
:start
IF EXIST "/absolute/path/lock.txt" (
    TYPE "/absolute/path/out.txt"
    DEL "/absolute/path/out.txt"
    DEL "/absolute/path/lock.txt"
) ELSE (
    ping -n 1
    GOTO start
)
```
worker.exe content:
```
#!/bin/bash
git "$@" > /absolute/path/out.txt
touch /absolute/path/lock.txt
```
![image](https://user-images.githubusercontent.com/35807926/80050327-8deabc80-8515-11ea-9113-5a6dac93806b.png)

**Comment by RussKie** [comment]
@mkbel this is so cool! 🚀

**Comment by mstv** [comment]
> Looks like GitExtensions is calling more git processes in parallel.

Yes, it does.

**Comment by brunoais** [comment]
Maybe this will work (not tested)


	git.bat content:

	@ECHO OFF
	execId=%RANDOM%
	worker.exe %execId% %*
	:start
	IF EXIST "/absolute/path/lock_%execId%.txt" (
		TYPE "/absolute/path/out_%execId%.txt"
		DEL "/absolute/path/out_%execId%.txt"
		DEL "/absolute/path/lock_%execId%.txt"
	) ELSE (
		ping -n 1
		GOTO start
	)

	worker.exe content:

	#!/bin/bash
	execId=$1
	shift
	git "$@" > "/absolute/path/out_$execId.txt"
	touch "/absolute/path/lock_$execId.txt"

**Comment by mkbel** [comment]
@brunoais 
I think it is good solution to allow git to be called several times in parallel.
After minor fix it works from wineconsole.
GitExtensions is still not happy with git input - maybe CRLF vs LF problem?
I created repository for this wrapper at https://github.com/mkbel/mkbel
And included your proposal.
You can either:
1. Do nothing about that.
1. Commit this change and create pull request to that repository - I will rewrite history then, so you will be clearly distinguished as contributor on github and your commit will include your email address, not just link to your github profile.
1. Ask me to not include your work.

**Comment by brunoais** [comment]
@mkbel Were you able to get any updates on this?

**Comment by mohamed-badaoui** [comment]
Distro: Debian10, Wine: 4.0
 
Installing 3.1 version of gitExtension, throw this error after adding git path.
Any update ? Thoughts ?

Thanks.


 
![image](https://user-images.githubusercontent.com/3098863/100707276-b3448080-33aa-11eb-999d-f3602a7a0ef3.png)

![image](https://user-images.githubusercontent.com/3098863/100707050-56e16100-33aa-11eb-80e9-1bde7096bce4.png)

**Comment by RussKie** [comment]
This relates to configuration of git remotes. The app expects fetch and
push urls come in pairs.

The condition was relaxed in one of the later versions, but I don't recall
in which.

**Comment by pmiossec** [comment]
@RussKie That's strange because he seems to already use the good version 3.1.1 ( https://github.com/gitextensions/gitextensions/releases/tag/v3.1.1 ) containing the fix (bug report #6562 & fix #6580 )

I will recommend to test a newer version 3.3.1 or 3.2.1 (@mohamed-badaoui why did you specifically tried this one?) and if that still not work, use the last officially supported unix version , v2.51...

@mohamed-badaoui Also, what is the git version you use? Perhaps too old?

**Comment by mohamed-badaoui** [comment]
@pmiossec Ok so I've succeed to make it work just upgrading wine from 4.x to wine 5.22

![image](https://user-images.githubusercontent.com/3098863/100717996-1a6a3100-33bb-11eb-97b6-47f32435cd15.png)

**Comment by Wirone** [comment]
@mohamed-badaoui I can't install `wine` nor `winehq` properly. Running GitExtensions on `wine` prompts for `wine32` and I can't install it because dependencies are not met (and when trying to install those packages, the other ones pop up). `winehq-stable` / `winehq-devel` can't be even installed because of unmet dependencies. Any ideas?

Ubuntu 20.10 with 5.9.12 kernel (but the same was with 20.04 and previous kernels).

**Comment by pmiossec** [comment]
@Wirone One idea like that: Did you tried to install them with `snap`?

**Comment by mohamed-badaoui** [comment]
@Wirone Sorry I can't help, my distro is Debian and to get wine 5.22, I just did :
`sudo apt install --install-recommends winehq-devel`

FYI: kernel is:
```
uname -a
Linux laptop 5.8.0-0.bpo.2-amd64 #1 SMP Debian 5.8.10-1~bpo10+1 (2020-09-26) x86_64 GNU/Linux
```
On Ubuntu did you try this ? 
http://ubuntuhandbook.org/index.php/2020/11/wine-5-22-released-install-in-ubuntu-20-10/

Once wine5.22, GitExtensions works well except for git commit :(  That's a little bit annoying for that kind of soft :')
I tried with Git for windows and cygwin.

**Comment by Wirone** [comment]
@mohamed-badaoui Yeah, I tried that article earlier, unfortunately it ends up with missing dependencies:

```
-> % sudo apt install --install-recommends winehq-stable wine-stable
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libfaudio0 : Depends: libavcodec58 (>= 7:4.2)
 libglib2.0-0 : Depends: libpcre3 but it is not installable
 libgnutls30 : Depends: libidn2-0 (>= 2.0.0) but it is not installable
 libgphoto2-6 : Depends: libgd3 (>= 2.1.0~alpha~) but it is not installable
                Depends: libxml2 (>= 2.7.4) but it is not installable
 libkrb5-3:i386 : Depends: libssl1.1:i386 (>= 1.1.0) but it is not installable
 libselinux1 : Depends: libpcre2-8-0 (>= 10.22) but it is not installable
 wine-stable-amd64 : Depends: libxml2 (>= 2.9.0) but it is not installable
                     Recommends: libsane or
                                 libsane1 but it is not going to be installed
                     Recommends: libtiff5 but it is not going to be installed
                     Recommends: libxslt1.1 but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

I saw [this](https://stackoverflow.com/a/62379887/842480) but I don't have such checkboxes in my Software & Updates window :sweat_smile: There must be some option or config that permits my `apt` from autoresolving and autoinstalling those dependencies :thinking:

**Comment by mohamed-badaoui** [comment]
@Wirone did you tried with aptitude? Else try to install these depedencies.

**Comment by Wirone** [comment]
@mohamed-badaoui Yeah, I tried `aptitude` (edited my previous comment in the meantime), no success. I think it's because it can't resolve `:i386` packages, because for example `libavcodec58` is installed on my system in correct version, but only for `amd64`:

```
-> % sudo apt-cache policy libavcodec58                                                                               
libavcodec58:
  Installed: 7:4.3.1-4ubuntu1
  Candidate: 7:4.3.1-4ubuntu1
  Version table:
 *** 7:4.3.1-4ubuntu1 500
        500 http://pl.archive.ubuntu.com/ubuntu groovy/universe amd64 Packages
        100 /var/lib/dpkg/status

-> % sudo apt-cache policy libavcodec58:i386
libavcodec58:i386:
  Installed: (none)
  Candidate: 7:4.3.1-4ubuntu1
  Version table:
     7:4.3.1-4ubuntu1 500
        500 http://pl.archive.ubuntu.com/ubuntu groovy/universe i386 Packages
```

However, resolving it by hand isn't possible probably:

```
-> % LANG=en_US.UTF-8 sudo apt install libavcodec58:i386 libssl1.1:i386 libglib2.0-0:i386 libgtk2.0-0:i386 libatk-wrapper-java-jni:i386
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package libatk-wrapper-java-jni:i386 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'libatk-wrapper-java-jni:i386' has no installation candidate
```

Too bad GitExtensions dropped Linux support, it is my favourite Git client (using 2.51.05 with Mono, but wanted to upgrade after I got email notification about [your comment](#issuecomment-736322671)) :slightly_frowning_face:

**Comment by mkbel** [comment]
@Wirone You need to enable multiarch in order to be able to install i386 packages from package manager. https://wiki.debian.org/Multiarch/Implementation#Using_multiarch

**Comment by Wirone** [comment]
@mkbel Of course I've added i386 architecture, still got problems. But it was generally hard time for my Ubuntu since after updating kernel I wasn't able to login properly, only Alt+F* and `startx` worked but without video and wifi drivers... So after fixing that with mainline kernel I removed architecture, `wine` and everything because I wasn't sure what caused my problems. Then I started again from scratch and still couldn't install and use Wine and GE.

**Comment by brunoais** [comment]
I'm happy to report that after a week of testing, that git extensions 3.5.4.12724 runs on wine 7.1 (later upgraded to wine 7.4) without any (functionality-wise) known issues.

- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.35.1.windows.2
- Microsoft Windows NT 6.1.7601.65536
- Mono 6.13.0 (tarball)
- DPI 96dpi (no scaling)
![image](https://user-images.githubusercontent.com/639467/158654348-36966c4d-3ff4-4b3f-a395-02e04c5dd83f.png)

There are, however, some visual glitches:

1. The main colors are off (may be fixable with some wine trick form winetricks)
2. When resizing the window, it flickers while trying to put all together (auto-fixed when resizing finishes)
3. Black bars when trying to load windows (fixed when whatever was loading finishes)

I was unable to have linux's git to run correctly, however, git.exe runs fine. The major issue with git.exe is it's performance. It is much, much slower than linux native git.

**Comment by mstv** [comment]
> I'm happy to report that after a week of testing, that git extensions 3.5.4.12724 runs on wine 7.1 (later upgraded to wine 7.4) without any (functionality-wise) known issues.

Thank you for sharing the great news.

> I was unable to have linux's git to run correctly.

Is GE unable to get the git output or unable to parse the output correctly?
Perhaps, GE's Git Command Log (hotkey `F12`) can help with the analysis - although it does not trace the output.

The [current dev build](https://ci.appveyor.com/project/gitextensions/gitextensions/branch/master/artifacts) is based on .NET 5 Core and runs pretty stable.
Does it work with Wine, too?

**Comment by brunoais** [comment]
> Is GE unable to get the git output or unable to parse the output correctly?

No. It's just that Wine whines it can't run ELF executables :rofl:... while I just want it to run natively in linux. Even though I [followed the manual](https://wiki.winehq.org/FAQ#How_do_I_launch_native_applications_from_a_Windows_application.3F). I asked about it but [no proper answers as of yet.](https://forum.winehq.org/viewtopic.php?f=2&t=36328)

I tried to make a translation script but I'm having a hard time telling wine that I want native git. Maybe I can figure out sth...

>  Does it work with Wine, too?

Yes. Almost out of the box. I had to do get wine 7.2 (there's no 7.4 64-bit, apparently). It doesn't work on wine 6.17, for me. It just throws an exception.
Also,
I had to install .net 5 by downloading it from microsoft's website and run it there. It was easy to know the URL, I only had to press the button and wine opened my local browser.

![image](https://user-images.githubusercontent.com/639467/158686621-9f35f688-d06c-4702-8f32-4a7ce2f868b0.png)

The only thing missing is for me to be able to run the linux git from gitextensions. The path can be translated with `winepath`. I'm fully sure native git would provide the performance I've been getting with GE 2.5

**Comment by gerhardol** [comment]
.NET 4 should be used for 3.5.4, may explain the visual glitches.

For Git support, you need something that behaves like git.exe in wine.
master has added support to handle WSL Git, so basically "wsl -d Ubuntu-20.04 git" is used instead of "git.exe".
That handling is hardcoded to paths in "\\wsl$", also some special handling of paths (translation back and forward).
Some similar wrapper for Git is needed and likely some special translation rules too.

**Comment by brunoais** [comment]
> .NET 4 should be used for 3.5.4, may explain the visual glitches.

I tried with both wine-mono and .NET 4.8 . The glitches while loading stay regardless.

> For Git support, you need something that behaves like git.exe in wine. master has added support to handle WSL Git, so basically "wsl -d Ubuntu-20.04 git" is used instead of "git.exe". That handling is hardcoded to paths in "\wsl$", also some special handling of paths (translation back and forward). Some similar wrapper for Git is needed and likely some special translation rules too.

From my perspective, I think I just need a way to call the native git executable. The path translation should be able to be taken care of by `winepath`.
I'm probably just missing some detail here in the process...
I'll see if I can figure out why I can't get an external program to run and return the output

**Comment by monoblaine** [comment]
I'm using Linux Mint 22 (Ubuntu 24.04) and running GitExtensions v4.2.1 using wine (v9.15) with "Git for Windows" still works, but there were always some imaginary "file mode changes" reported on commit windows (I didn't want to use `git config core.fileMode=false`) and they were quite disturbing.

I really, really wanted to use the native git so I've tried @mkbel's approach (https://github.com/mkbel/mkbel) with no success. I did some digging and found out why it failed:

1. Using `%RANDOM%` for the name of `out` and `lock` files does not work because GitExtensions sometimes executes git commands in parallel and [according to this article](https://web.archive.org/web/20190216195619/https://blogs.msdn.microsoft.com/oldnewthing/20100617-00/?p=13673) Windows cmd uses `time` as a seed value for `srand`, and `time` has a resolution of only one second which leads to duplicate `%RANDOM%` values and this makes some of the `out` files get rewritten during parallel command execution.
2. The `type` command messes up the encoding of files.

I fixed the first issue by replacing `%RANDOM%` with `uuidgen.exe` copied from a Windows machine.

I didn't really have time to do enough research for the second issue so I simply used `cat.exe` which is created during Git for Windows setup. (`C:\Program Files\Git\usr\bin\cat.exe`)

Below is the edited script:

```bat
@echo off

:: TODO: Edit this path.
set git_dir=Z:\home\serhan\Projects\mkbel
for /F "tokens=*" %%g in ('%git_dir%/uuidgen.exe') do (set execId=%%g)
%git_dir%\worker.sh %execId% %*

:start
if exist "%git_dir%\lock_%execId%.txt" (
    "C:\Program Files\Git\usr\bin\cat.exe" "%git_dir%\out_%execId%.txt"
    del "%git_dir%\out_%execId%.txt"
    del "%git_dir%\lock_%execId%.txt"
) else (
    ping -n 1 > nul 2>&1
    goto start
)
```

![Screenshot_2024-08-20_21-24-56](https://github.com/user-attachments/assets/87327df3-b59a-4da1-8403-574030f1f7f4)

**Comment by gerhardol** [comment]
> I really, really wanted to use the native git so I've tried @mkbel's approach (https://github.com/mkbel/mkbel) with no success. I did some digging and found out why it failed:

You may be able to adopt the WSL mode to work with Linux Git. GE has still some checks of the environment it runs in, may be possible to detect wine for special handling.

Is there anything that should be documented to get started for those that want to try this?

**Comment by monoblaine** [comment]
Well, it appears it was too soon to get excited because my solution above failed miserably on some other parts like staging that needs piping data into the git executable.

In a bash script you can write something like the following in order detect whether piped input exists or not:

```sh
if [ -t 0 ]; then
    echo "nothing's piped"
else
    while read -r line ; do
        echo $line
    done
fi
```

I was unable to find something similar to use in the Windows batch script. I guess that's it for now, I'm giving up.

@gerhardol I'd really appreciate some heads-up on the WSL mode.

**Comment by gerhardol** [comment]
If a repo is detected as WSL (\\wsl$\ or \\wsl.localhost\) the WSL git executable is used instead by using "wsl.exe -d Ubuntu git <args>". There is some special handling of paths, whether the windows Git or wsl Git executable is used.
In the Wine setup the Git executable could be handled in a similar way if that simplifies the handling.

**Comment by monoblaine** [comment]
Thank you @gerhardol, I'll definitely look into it sometime soon. In the meantime I've totally succumbed to my obsession and was able to create a .NET Framework console application that is more promising than the Windows batch file:

https://github.com/monoblaine/wine_git

For some reason it's still slow; but I can do staging, partial staging and create/list commits. That's all I want for now.

**Comment by pmiossec** [comment]
> In the meantime I've totally succumbed to my obsession and was able to create a .NET Framework console application that 
> For some reason it's still slow; 

Maybe you could try to convert it to a "dotnet core console app using native aot" to get better performance.
https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Clinux-ubuntu

**Comment by monoblaine** [comment]
Thanks for the great tip @pmiossec, it made a huge difference! The performance improvement is far beyond my expectations!

Here is the release for those interested:

https://github.com/monoblaine/wine_git/releases/tag/v0.2.1

(After hitting the keys randomly like a golden retriever for a period of time I managed to automatically create a release with artifacts.)

**Comment by gerhardol** [comment]
@monoblaine If wine+wine_git is working well, could you describe this in https://github.com/gitextensions/gitextensions/wiki/How-To%3A-run-Git-Extensions-on-Linux and the main README?

**Comment by jens-br** [comment]
> [@monoblaine](https://github.com/monoblaine) If wine+wine_git is working well, could you describe this in https://github.com/gitextensions/gitextensions/wiki/How-To%3A-run-Git-Extensions-on-Linux and the main README?

Some intuitive bullet points for the required steps would be extremely helpful, since I have to admit that I somewhat lost track of what needs to be done.

**Comment by gerhardol** [comment]
> How could this be done? 

wine is for real Linux, I do not think wine is available at all in WSL.
In that case the binary would run in Linux instead of Win but with a Git executable that runs outside wine.

For your usecase you will need to adapt GE itself.

**Comment by skdsp** [comment]
@gerhardol At the end we also need it for a native Linux and I am also running GE 5.0 there with Wine RC and .Net 8.0.0...we really would like to use GE in both worlds. 
So could you give me a hint about the intended WSL workaround?

**Comment by monoblaine** [comment]
> [@monoblaine](https://github.com/monoblaine) If wine+wine_git is working well, could you describe this in https://github.com/gitextensions/gitextensions/wiki/How-To%3A-run-Git-Extensions-on-Linux and the main README?

Hey @gerhardol, sorry for the late reply. I'm fighting the covid at the moment, but I'll be happy to document the required steps and send a PR hopefully next week. But I'll have to test GE v5 before that.  In v4.2 (the version that I'm currently using), there were some issues with Wine's handling of `RichTextBox` content (like new-line characters getting ignored). If they are still present, we may have to change some of the things in the GE source. (https://github.com/monoblaine/gitextensions/commit/685a826a6637380138f3de041c26aa40d3c822ec is an example)

> An automatic wine detection would be very nice for the future.

@skdsp [this](https://github.com/monoblaine/gitextensions/commit/09db2cad6f2efae92dc0bc7cb4e39fa734c38ad9) is how I make sure if GE is running under Wine or not. Should've added some reference to the original code though, I probably copied it from StackOverflow.

**Comment by monoblaine** [comment]
I've created a new release for `wine_git` here:

https://github.com/monoblaine/wine_git/releases/tag/v0.4.0

I did some testing on Arch Linux with Xfce using the portable version of GitExtensions v5.2.1. Everything (besides the `RichTextBox` issue I've mentioned before) looks OK so far. Additional tests would be appreciated before I send a PR regarding setup, etc.

![Image](https://github.com/user-attachments/assets/5bd06f3f-25c4-4bf7-853d-64de7377f46d)

Here are the steps I followed:

Install `wine`.

```sh
$ wine --version
wine-10.3
```

Add `export WINEDEBUG=-all` to your `~/.bashrc` in order to prevent Wine from spitting weird debug messages.

After installing `wine` download `wine_git` and create a folder for its artifacts:

```sh
mkdir -p ~/wine/tmp
cp wine_git.exe worker.sh settings.ini ~/wine/
```

Update `path_to_wine_git_folder` in `settings.ini`:

```ini
# Make sure the path starts with "Z:"
path_to_wine_git_folder="Z:/home/serhan/wine"
```
Update `path_to_wine_git_folder` in `worker.sh`:

```sh
# No "Z:" prefix needed here.
path_to_wine_git_folder="/home/serhan/wine"
```

Now open a terminal and `cd` to a Git repository.

The following command should work fine:

```sh
$ wine ~/wine/wine_git.exe status
```

Example:

```
[serhan@SD-ARCHVM gitextensions]$ wine ~/wine/wine_git.exe status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

Now download GitExtensions portable version and open it (`GitExtensions.exe`). It will prompt you to install ".NET 8.0 Desktop Runtime". Do it.

When GitExtensions is open, configure your "git.exe". The file picker demands a file name that starts with `git` so enter the path to `wine_git.exe` manually instead:

![Image](https://github.com/user-attachments/assets/40819692-4d6b-40f6-8fc3-a6d3e277cf6b)

**Comment by RussKie** [comment]
> I've created a new release for `wine_git` here:

Thank you for sharing.

Should we replace content of https://github.com/gitextensions/gitextensions/wiki/How-To%3A-run-Git-Extensions-on-Linux with your instructions?

**Comment by monoblaine** [comment]
> Should we replace content of https://github.com/gitextensions/gitextensions/wiki/How-To%3A-run-Git-Extensions-on-Linux with your instructions?

Oh, I didn't know it was already put on the record! 😅 The steps I've given were tested on a newly installed virtual machine, so please feel free to do so.

**Comment by monoblaine** [comment]
Haha it even works on macOS! 😂 (an Intel-based Hackintosh® VM, actually)

```sh
brew install --cask --no-quarantine wine-stable
echo -e "\nexport WINEDEBUG=-all\nexport MVK_CONFIG_LOG_LEVEL=0" >>~/.zshrc
source ~/.zshrc
wine <path to GitExtensions>/GitExtensions.exe
```

But it appears Wine on macOS has problems with `Console.OpenStandardInput()` if there's in fact no input really redirected (?). I had to modify `wine_git` so that it waits no longer than 100ms until the first few bytes are read:

```diff
                 using var inputStream = Console.OpenStandardInput();
                 using var redirectedInput = File.OpenWrite(pathToRedirectedInput!);
-                inputStream.CopyTo(redirectedInput);
-                isInputReallyRedirected = redirectedInput.Length > 0;
+                using var cts = new CancellationTokenSource(100);
+                var ct = cts.Token;
+                var buffer = new Byte[4096];
+                var firstBytesReadTask = inputStream.ReadAsync(buffer, 0, buffer.Length, ct);
+                try {
+                    firstBytesReadTask.Wait(ct);
+                    var bytesRead = firstBytesReadTask.Result;
+                    if (bytesRead > 0) {
+                        redirectedInput.Write(buffer, 0, bytesRead);
+                    }
+                    while ((bytesRead = inputStream.Read(buffer, 0, buffer.Length)) > 0) {
+                        redirectedInput.Write(buffer, 0, bytesRead);
+                    }
+                    isInputReallyRedirected = redirectedInput.Length > 0;
+                }
+                catch (OperationCanceledException) {
+                    isInputReallyRedirected = false;
+                    Log("read stdin timed out.");
+                }
             }
             if (!isInputReallyRedirected) {
                 File.Delete(pathToRedirectedInput!);
```

I can publish a new release if someone likes to try this out.

![Image](https://github.com/user-attachments/assets/11a13157-f69a-4ec7-b258-b9508367416e)

***

**Up next:** *How I ran GitExtensions on Windows 95*

**Comment by gerhardol** [comment]
Great!
Were any changes required to GitExtensions binary? (or should something be done?)

And it should be possible to use native git.exe in a similar way as for WSL. But it affects many source lines in GE still.
I am not using native Linux myself, you need this setup and have the time and interest to implement such a change.

**Comment by monoblaine** [comment]
> Were any changes required to GitExtensions binary?

No, I just downloaded the latest portable version both on Linux and macOS. I don't have time to do extensive testing but the following things (things I primarly use GE for) work fine:

* File staging
* Partial staging
* Creating commits
* Viewing diffs in commits

**Comment by brunoais** [comment]
@gerhardol I've been trying in parallel to do this more seamless by doing some changes to GitExtensions. I've been in and out around that since 2023, tbh 😅. I just didn't want to bother you with linux bcs this is not a linux project.
My issues with monoblaine solution is that all paths and such are in windows notation (although they work) and features such as opening in the file explorer are annoyingly opening the lackluster and barebones wine file explorer.

For now, I'm trying to figure out if I can get away with only editing `OsShellUtil.cs` and `PathUtil.cs`. I may be able to trick GitExtensions to believe everything is WSL and so it should just assume WSL.

I'm also aiming to target flatpak, so I'm trying to get it to use `/usr/bin/flatpak-spawn --host` (only changed some code for now... No idea if it will work).

If you don't know any of what I just babbled about because it's not windows, don't worry about how to make it work in linux. That's for me to figure out. What I need is only to know where I'd need to change so I can use linux path notations and override running any external programs such as opening explorer where a file is or copying a file path.

In parallel, I'm trying to figure out ways to compile the code (using the same service you do is a very slow process when doing development) 


tl;dr: I only need to know where to edit so GitExtensions can handle linux paths from the user's perspective, even if it receives and translates just to translate back when acting on the system (or running git).

**Comment by monoblaine** [comment]
> [...] and features such as opening in the file explorer are annoyingly opening the lackluster and barebones wine file explorer.

Hey @brunoais, if what you need is the "Show in folder" feature it could be accomplished by a few steps:

```sh
user_name=$(whoami)
echo -e "@echo off\n\n/home/$user_name/wine_util/wine_explorer.sh %*" >~/.wine/drive_c/windows/wine_explorer.bat
mkdir ~/wine_util
echo -e '#!/bin/bash

if [[ "$1" == "/select," ]]; then
    shift
fi

path=${1//\\/\/}
path=${path/Z:/}

# TODO: Replace thunar with your file explorer
thunar "$path"' >~/wine_util/wine_explorer.sh
```

Then here,

https://github.com/gitextensions/gitextensions/blob/14506796ac8d998327b6fb607a26c116a4a2da3e/src/app/GitCommands/Git/OsShellUtil.cs#L32-L34

Making the following change will be enough:

```diff
-        public static void OpenWithFileExplorer(string arguments, bool quote = true) => new Executable("explorer.exe").Start(quote ? arguments.Quote() : arguments);
+        public static void OpenWithFileExplorer(string arguments, bool quote = true) => new Executable("wine_explorer.bat").Start(quote ? arguments.Quote() : arguments);
```

**Comment by brunoais** [comment]
@monoblaine I already went beyond that although I didn't test this stuff yet.

![Image](https://github.com/user-attachments/assets/c8eb37ac-f59f-460a-a68d-4e34e7f10f5f)

But what I'm after is an experience closer to version 2.5 (including performance) without it crashing or needing a resize to refresh the calculations. Wine allows running linux executables directly from it, so windows executables/runtimes can execute separate linux compiled programs without the need for an intermediary (it just takes an adjustment to the registry).
If I can run this in a normalized platform such as using flatpak, I can make it much easier for me to setup and package whenever I want to update. Maybe others can also benefit from it.

That's my rationale, anyway.

**Comment by gerhardol** [comment]
> **tl;dr:** I only need to know where to edit so GitExtensions can handle linux paths from the user's perspective, even if it receives and translates just to translate back when acting on the system (or running git).

wine_git.exe is kind of similar to https://github.com/andy-5/wslgit
Many callsites where Git executables need to be selected and (file) in/out has to be converted is handled, is very similar with with ge-on-wine. App (win/wine) paths <-> wsl/linux paths.
But Win executables can still be executed in WSL, so there are some additions.

Some of the special handling (and having WSL Git as a second class implementation is improved (like #12264).
wine support adds to the complexity, this must be done with care...

**Comment by RussKie** [comment]
Folks, this issue is closed, I suggest opening a draft pull-request and continue the discussions there.
